compiler/front/nimconf.nim

```diff
@@ -17,29 +17,161 @@ import
   ],
   compiler/front/[
     commands,
-    msgs,
     options,
     scriptconfig
   ],
   compiler/ast/[
     lexer,
-    reports,
     idents,
     wordrecg,
     llstream,
+    lineinfos,
     ast
   ],
   compiler/utils/[
     pathutils,
+    idioms,
   ]
 
+type
+  ConfigEventKind* = enum
+    ## events/errors arising from parsing and processing a compiler config file
+
+    # fatal errors begin
+    cekInternalError
+    # fatal errors end
+
+    # users errors begin
+
+    # lexer generated
+    cekLexerErrorDiag        ## lexer error being forwarded
+
+    # expected token
+    cekParseExpectedX        ## expected some token
+    cekParseExpectedCloseX   ## expected closing ')', ']', etc
+    cekParseExpectedIdent    ## expected an identifier
+
+    # invalid input
+    cekInvalidDirective
+    # user errors end
+
+    # warning begin
+    cekLexerWarningDiag      ## warning from the lexer
+    # warning end
+
+    # hint begin
+    cekLexerHintDiag         ## hint from the lexer
+    # hint end
+
+    # user output start
+    cekWriteConfig           ## write out the config
+    # user output end
+
+    # debug start
+    cekDebugTrace            ## config debug trace
+    cekDebugReadStart        ## start config file read
+    cekDebugReadStop         ## stop config file read
+    # debug end
+
+    # progress begin
+    cekProgressConfStart
+    # progress end
+
+  ConfigFileEvent* = object
+    case kind*: ConfigEventKind:
+      of cekParseExpectedX, cekParseExpectedCloseX, cekParseExpectedIdent,
+         cekInvalidDirective, cekWriteConfig, cekDebugTrace:
+        location*: TLineInfo         ## diagnostic location
+      of cekInternalError, cekLexerErrorDiag, cekLexerWarningDiag,
+         cekLexerHintDiag:
+        lexerDiag*: LexerDiag
+      of cekDebugReadStart, cekDebugReadStop, cekProgressConfStart:
+        discard
+    instLoc*: InstantiationInfo ## instantiation in lexer's source
+    msg*: string
+  
+  NimConfEvtHandler* = proc(config: ConfigRef,
+                            evt: ConfigFileEvent,
+                            reportFrom: InstantiationInfo,
+                            eh: TErrorHandling = doNothing): void
+
+var
+  cfgEvtHandler: NimConfEvtHandler
+    ## event handler ensure this is set when a public proc that can call the
+    ## `handle*` templates is invoked (take it in as a param). Currently, this
+    ## is only set by `loadConfig`. This allows the module to not depend upon
+    ## `reports` infrastructure which is being refactored/simplified
+  lexDiagOffset = 0
```

Comment:
As an alternative to the use of globals here, you could introduce a separate `NimConfParser` (or similar) that would look like this:
```nim
type NimConfParser = object
  lexer: Lexer
  condStack: seq[bool]
  config: ConfigRef

  cfgEvtHandler: NimConfEvtHandler
  lexDiagOffset: int
```

and then pass an instance of this object to each `parseX` procedure.

Also, it seems like `lexDiagOffset` is only ever set `0`?

Reply:
Yeah, I looked at this briefly, it does get a bit weird with the "output" `ConfigRef` vs the one on the lexer, IIRC. I'll give it another go.